From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shahbaz Youssefi <syoussefi@chromium.org>
Date: Tue, 9 Jun 2026 13:58:24 -0400
Subject: Vulkan: Fix advanced blend validation + backend skip

Advanced blend is only valid when used with draw buffer 0.  When
enabled, non-0 draw buffers shouldn't be enabled.

In the Vulkan backend, blend is skipped for disabled draw buffers
because Vulkan's validation is less forgiving to mixing blend and
advanced blend for disabled attachments (See
VUID-VkPipelineColorBlendAttachmentState-advancedBlendIndependentBlend-01407)

Bug: chromium:513257423
Change-Id: I9671015fef16bcb7b170b182905d481c37c2371d
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/7914802
Reviewed-by: Amirali Abdolrashidi <abdolrashidi@google.com>
Commit-Queue: Shahbaz Youssefi <syoussefi@chromium.org>

diff --git a/src/libANGLE/ErrorStrings.h b/src/libANGLE/ErrorStrings.h
index 4d8bae281200214d36abb52f54ecc3ae70aa901a..d33a936c6f1fa7339b5ab04c193eb67cafe3ebb4 100644
--- a/src/libANGLE/ErrorStrings.h
+++ b/src/libANGLE/ErrorStrings.h
@@ -17,7 +17,7 @@ namespace err
 // clang-format off
 inline constexpr const char *k3DDepthStencil = "Format cannot be GL_DEPTH_COMPONENT or GL_DEPTH_STENCIL if target is GL_TEXTURE_3D.";
 inline constexpr const char *kANGLECopyTextureMissingRequiredExtension = "Copy*TextureCHROMIUM from EXTERNAL_OES to integer format requires OES_EGL_image_external_essl3.";
-inline constexpr const char *kAdvancedBlendEquationWithMRT = "Advanced blend equation can only be used when only one draw buffer is not NONE.";
+inline constexpr const char *kAdvancedBlendEquationWithMRT = "Advanced blend equations are used with a draw buffer other than draw buffer zero or with multiple draw buffers.";
 inline constexpr const char *kAtomicCounterResourceName = "Active atomic counter resources are not assigned name strings.";
 inline constexpr const char *kAttributeListNotNull = "Attribute list must be NULL or GL_NONE.";
 inline constexpr const char *kAttributeNameNull = "Attribute name is null.";
diff --git a/src/libANGLE/renderer/vulkan/ContextVk.cpp b/src/libANGLE/renderer/vulkan/ContextVk.cpp
index fac0b185f6b59dca4ebde755ff8c6425273170d6..07cc9b792fb9ea203cc22f94c14d4e29cc8e8700 100644
--- a/src/libANGLE/renderer/vulkan/ContextVk.cpp
+++ b/src/libANGLE/renderer/vulkan/ContextVk.cpp
@@ -4782,6 +4782,18 @@ void ContextVk::updateMissingAttachments()
     }
 }
 
+void ContextVk::updateBlendEnabled()
+{
+    const gl::DrawBufferMask framebufferMask = mState.getDrawFramebuffer()->getDrawBufferMask();
+    const gl::DrawBufferMask enabledBlend    = mState.getBlendStateExt().getEnabledMask();
+
+    // Filter out blend for disabled attachments.  If advanced blend is enabled, Vulkan
+    // forbids blend from being specified on the other attachments (same as GL, but GL
+    // ignores blend on disabled attachments).
+    mGraphicsPipelineDesc->updateBlendEnabled(&mGraphicsPipelineTransition,
+                                              enabledBlend & framebufferMask);
+}
+
 void ContextVk::updateBlendFuncsAndEquations()
 {
     const gl::BlendStateExt &blendStateExt = mState.getBlendStateExt();
@@ -5441,8 +5453,7 @@ angle::Result ContextVk::syncState(const gl::Context *context,
                 updateDepthRange(glState.getNearPlane(), glState.getFarPlane());
                 break;
             case gl::state::DIRTY_BIT_BLEND_ENABLED:
-                mGraphicsPipelineDesc->updateBlendEnabled(
-                    &mGraphicsPipelineTransition, glState.getBlendStateExt().getEnabledMask());
+                updateBlendEnabled();
                 updateDither();
                 updateAdvancedBlendEquations(programExecutable);
                 break;
@@ -5706,6 +5717,7 @@ angle::Result ContextVk::syncState(const gl::Context *context,
                                glState.getFarPlane());
                 updateColorMasks();
                 updateMissingAttachments();
+                updateBlendEnabled();
                 updateRasterizationSamples(drawFramebufferVk->getSamples());
                 updateRasterizerDiscardEnabled(
                     mState.isQueryActive(gl::QueryType::PrimitivesGenerated));
@@ -6482,6 +6494,7 @@ angle::Result ContextVk::onFramebufferChange(FramebufferVk *framebufferVk, gl::C
 
     // Attachments might have changed.
     updateMissingAttachments();
+    updateBlendEnabled();
 
     if (mState.getProgramExecutable())
     {
diff --git a/src/libANGLE/renderer/vulkan/ContextVk.h b/src/libANGLE/renderer/vulkan/ContextVk.h
index 733cf2bd5689a0327520d69f9637cf2d5c7ab2c4..5c8ad5ee02949993b8b1e44997b8f34e5cdd3f1e 100644
--- a/src/libANGLE/renderer/vulkan/ContextVk.h
+++ b/src/libANGLE/renderer/vulkan/ContextVk.h
@@ -492,6 +492,7 @@ class ContextVk : public ContextImpl, public vk::Context, public MultisampleText
                                        gl::SamplerFormat format,
                                        gl::Texture **textureOut);
     void updateColorMasks();
+    void updateBlendEnabled();
     void updateBlendFuncsAndEquations();
 
     void handleError(VkResult errorCode,
diff --git a/src/libANGLE/renderer/vulkan/FramebufferVk.cpp b/src/libANGLE/renderer/vulkan/FramebufferVk.cpp
index 607fecbccf3d4c09ded359e084749e77d5f61f79..c368f1c0d5a126a380c22f6b813247f3b9001a6a 100644
--- a/src/libANGLE/renderer/vulkan/FramebufferVk.cpp
+++ b/src/libANGLE/renderer/vulkan/FramebufferVk.cpp
@@ -2759,6 +2759,7 @@ angle::Result FramebufferVk::syncState(const gl::Context *context,
     if (shouldUpdateColorMaskAndBlend)
     {
         contextVk->updateColorMasks();
+        contextVk->updateBlendEnabled();
         contextVk->updateBlendFuncsAndEquations();
     }
 
diff --git a/src/libANGLE/validationES.cpp b/src/libANGLE/validationES.cpp
index 3eeb53c72b3999d0838ba2d1ad82b88840f0c726..35be337ac14fd1b139ca5f5924654110af343501 100644
--- a/src/libANGLE/validationES.cpp
+++ b/src/libANGLE/validationES.cpp
@@ -4164,28 +4164,27 @@ const char *ValidateDrawStates(const Context *context, GLenum *outErrorCode)
         }
     }
 
-    // Advanced blend equation can only be enabled for a single render target.
+    // Advanced blend equation can only be enabled for color attachment zero and with only that
+    // attachment enabled.
     const BlendStateExt &blendStateExt = state.getBlendStateExt();
-    if (ANGLE_UNLIKELY(blendStateExt.getUsesAdvancedBlendEquationMask().any()))
+    const gl::DrawBufferMask drawBuffersWithAdvancedBlend =
+        blendStateExt.getUsesAdvancedBlendEquationMask() & blendStateExt.getEnabledMask();
+    if (ANGLE_UNLIKELY(drawBuffersWithAdvancedBlend.any()))
     {
+        // Note: Framebuffer::getDrawBufferMask() excludes non-existing draw buffers.  But if
+        // glDrawBuffers sets them to non-NONE, draw with advanced blend is still expected to fail.
+        // So we can't do a quick check with getDrawBufferMask() alone.
         const size_t drawBufferCount            = framebuffer->getDrawbufferStateCount();
-        uint32_t advancedBlendRenderTargetCount = 0;
-
-        for (size_t drawBufferIndex : blendStateExt.getUsesAdvancedBlendEquationMask())
+        const bool advancedBlendOnDrawBufferZero =
+            drawBuffersWithAdvancedBlend[0] && framebuffer->getDrawBufferState(0) != GL_NONE;
+        for (size_t drawBufferIndex = 1; drawBufferIndex < drawBufferCount; ++drawBufferIndex)
         {
-            if (drawBufferIndex < drawBufferCount &&
-                framebuffer->getDrawBufferState(drawBufferIndex) != GL_NONE &&
-                blendStateExt.getEnabledMask().test(drawBufferIndex) &&
-                blendStateExt.getUsesAdvancedBlendEquationMask().test(drawBufferIndex))
+            if (framebuffer->getDrawBufferState(drawBufferIndex) != GL_NONE &&
+                (advancedBlendOnDrawBufferZero || drawBuffersWithAdvancedBlend[drawBufferIndex]))
             {
-                ++advancedBlendRenderTargetCount;
+                return kAdvancedBlendEquationWithMRT;
             }
         }
-
-        if (advancedBlendRenderTargetCount > 1)
-        {
-            return kAdvancedBlendEquationWithMRT;
-        }
     }
 
     // Dual-source blending functions limit the number of supported draw buffers.
diff --git a/src/tests/gl_tests/AdvancedBlendTest.cpp b/src/tests/gl_tests/AdvancedBlendTest.cpp
index 15a6c496b30c10e0eb6b6cf0e0d82181e114dbd2..bc79347aef7f11662a2a08b0d08182cd0ac5a33d 100644
--- a/src/tests/gl_tests/AdvancedBlendTest.cpp
+++ b/src/tests/gl_tests/AdvancedBlendTest.cpp
@@ -34,6 +34,129 @@ class AdvancedBlendTest : public ANGLETest<>
 class AdvancedBlendTestES32 : public AdvancedBlendTest
 {};
 
+// Test that advanced blend cannot be used when a non-zero draw buffer is enabled.
+TEST_P(AdvancedBlendTest, NonZeroDrawBufferDisallowed)
+{
+    ANGLE_SKIP_TEST_IF(!EnsureGLExtensionEnabled("GL_KHR_blend_equation_advanced"));
+
+    GLFramebuffer fbo;
+    glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+
+    std::array<GLRenderbuffer, 4> rbo;
+    for (uint32_t i = 0; i < 4; ++i)
+    {
+        glBindRenderbuffer(GL_RENDERBUFFER, rbo[i]);
+        glRenderbufferStorage(GL_RENDERBUFFER, GL_RGBA8, 1, 1);
+        glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + i, GL_RENDERBUFFER,
+                                  rbo[i]);
+    }
+    EXPECT_GL_FRAMEBUFFER_COMPLETE(GL_FRAMEBUFFER);
+    ASSERT_GL_NO_ERROR();
+
+    GLenum enabled[4] = {GL_COLOR_ATTACHMENT0, GL_COLOR_ATTACHMENT1, GL_COLOR_ATTACHMENT2,
+                         GL_COLOR_ATTACHMENT3};
+    glDrawBuffers(4, enabled);
+
+    constexpr char kVS[] = R"(#version 300 es
+precision highp float;
+void main()
+{
+    switch (gl_VertexID)
+    {
+        case 0:  gl_Position = vec4(-1, -1, 0, 1); break;
+        case 1:  gl_Position = vec4( 3, -1, 0, 1); break;
+        default: gl_Position = vec4(-1, 3, 0, 1); break;
+    }
+})";
+
+    constexpr char kFS[] = R"(#version 300 es
+#extension GL_KHR_blend_equation_advanced : require
+layout (blend_support_multiply) out;
+precision mediump float;
+out vec4 fragColor;
+void main()
+{
+    fragColor = vec4(1., 0., 0., 1.);
+})";
+    ANGLE_GL_PROGRAM(program, kVS, kFS);
+    glUseProgram(program);
+
+    glEnable(GL_BLEND);
+    glBlendEquation(GL_MULTIPLY_KHR);
+
+    // Invalid draw because all 4 draw buffers are enabled.
+    ASSERT_GL_NO_ERROR();
+    glDrawArrays(GL_TRIANGLES, 0, 3);
+    EXPECT_GL_ERROR(GL_INVALID_OPERATION);
+
+    enabled[1] = GL_NONE;
+    enabled[2] = GL_NONE;
+    glDrawBuffers(4, enabled);
+
+    // Still invalid because draw buffer #3 is enabled.
+    ASSERT_GL_NO_ERROR();
+    glDrawArrays(GL_TRIANGLES, 0, 3);
+    EXPECT_GL_ERROR(GL_INVALID_OPERATION);
+
+    enabled[0] = GL_NONE;
+    enabled[1] = GL_COLOR_ATTACHMENT1;
+    enabled[3] = GL_NONE;
+    glDrawBuffers(4, enabled);
+
+    // Even though a single draw buffer is enabled, it's not index 0 so it's still invalid
+    ASSERT_GL_NO_ERROR();
+    glDrawArrays(GL_TRIANGLES, 0, 3);
+    EXPECT_GL_ERROR(GL_INVALID_OPERATION);
+
+    enabled[0] = GL_COLOR_ATTACHMENT0;
+    enabled[1] = GL_NONE;
+    glDrawBuffers(4, enabled);
+
+    // When only color attachment 0 is enabled, the draw is valid.
+    glClearColor(1, 1, 1, 1);
+    glClear(GL_COLOR_BUFFER_BIT);
+    glDrawArrays(GL_TRIANGLES, 0, 3);
+    ASSERT_GL_NO_ERROR();
+    EXPECT_PIXEL_COLOR_EQ(0, 0, GLColor::red);
+
+    // It's also ok to render to multiple attachments if blend is not enabled on the attachment that
+    // has advanced blend..
+    if (EnsureGLExtensionEnabled("GL_OES_draw_buffers_indexed"))
+    {
+        // Enable two attachments
+        enabled[1] = GL_COLOR_ATTACHMENT1;
+        glDrawBuffers(4, enabled);
+
+        // Enable blend only on attachment 0
+        glDisable(GL_BLEND);
+        glEnableiOES(GL_BLEND, 0);
+        glDisableiOES(GL_BLEND, 1);
+
+        // Set advanced blend on the attachment that has blend disabled.  Set a non-advanced blend
+        // on the one that has blend enabled.
+        glBlendEquationiOES(0, GL_FUNC_ADD);
+        glBlendFunciOES(0, GL_ONE, GL_ONE);
+        glBlendEquationiOES(1, GL_MULTIPLY_KHR);
+
+        glClearColor(0, 1, 0, 1);
+        glClear(GL_COLOR_BUFFER_BIT);
+        glDrawArrays(GL_TRIANGLES, 0, 3);
+        ASSERT_GL_NO_ERROR();
+        EXPECT_PIXEL_COLOR_EQ(0, 0, GLColor::yellow);
+
+        // If advanced blend is enabled on a NONE attachment, that's also ok.
+        enabled[1] = GL_NONE;
+        glDrawBuffers(4, enabled);
+        glEnableiOES(GL_BLEND, 1);
+
+        glClearColor(0, 0, 1, 1);
+        glClear(GL_COLOR_BUFFER_BIT);
+        glDrawArrays(GL_TRIANGLES, 0, 3);
+        ASSERT_GL_NO_ERROR();
+        EXPECT_PIXEL_COLOR_EQ(0, 0, GLColor::magenta);
+    }
+}
+
 void AdvancedBlendTest::callBlendBarrier(APIExtensionVersion usedExtension)
 {
     ASSERT(usedExtension == APIExtensionVersion::Core || usedExtension == APIExtensionVersion::KHR);
